### PR TITLE
[coop] Convert Thread.ByteArrayToCurrentDomain and Thread.ByteArrayToRootDomain.

### PR DESCRIPTION
**Pull request by jaykrell**

### CODE REVIEW

File: mono/metadata/threads.c [hunk 1]
```diff
@@ -1740,39 +1740,56 @@ ves_icall_System_Threading_Thread_SetPriority (MonoThreadObjectHandle this_obj,
...
-	if (mono_object_domain (arr) == domain)
+	if (MONO_HANDLE_DOMAIN (arr) == domain)
 		return arr;
 
-	copy = mono_array_new_checked (domain, mono_defaults.byte_class, arr->max_length, error);
-	memmove (mono_array_addr (copy, guint8, 0), mono_array_addr (arr, guint8, 0), arr->max_length);
-	return copy;
+	size_t const size = mono_array_handle_length (arr);
+
+	// Capture arrays into array for looping.
```
Comment (luhenry) [inline mono/metadata/threads.c]: Couldn't you just use 2 local variables here? The loop really obfuscate what you are trying to do for no apparent reason, there are only 2 elements and AFAICT there isn't going to be a third one.

Comment (jaykrell) [reply mono/metadata/threads.c]: 2 variables, and then 2 pins, and 2 unpins, prefer that? It seems clear with loops.

Comment (luhenry) [reply mono/metadata/threads.c]: It isn't clear to me who didn't write the code :wink:

Comment (jaykrell) [reply mono/metadata/threads.c]: Even with the comment?

File: mono/metadata/threads.c [hunk 2]
```diff
@@ -1740,39 +1740,56 @@ ves_icall_System_Threading_Thread_SetPriority (MonoThreadObjectHandle this_obj,
...
+	};
+	goto_if_nok (error, exit);
+
+	guint i;
+
+	// Pin both arrays.
+	for (i = 0; i <= 1; ++i)
+		arrays [i].p = mono_array_handle_pin_with_size (arrays [i].handle, size, 0, &arrays [i].gchandle);
+	goto_if_nok (error, exit);
+
```
Comment (jaykrell) [inline mono/metadata/threads.c]: hah and this line is extra too -- hopefully can't run out of pinning handles..

File: mono/metadata/threads.c [hunk 3]
```diff
@@ -1740,39 +1740,52 @@ ves_icall_System_Threading_Thread_SetPriority (MonoThreadObjectHandle this_obj,
...
+	HANDLE_FUNCTION_ENTER ()
 
-	error_init (error);
-	if (!arr)
-		return NULL;
+	if (!arr || MONO_HANDLE_IS_NULL (arr))
+		return MONO_HANDLE_NEW (MonoArray, NULL);
 
-	if (mono_object_domain (arr) == domain)
+	if (MONO_HANDLE_DOMAIN (arr) == domain)
```
Comment (jaykrell) [inline mono/metadata/threads.c]: It is a little funny to return a mutable alias if the domains happen to be equal. Faster, but could confuse caller. But that's what it was doing.

**Comment by jaykrell** [comment]
review

File: mono/metadata/threads.c [hunk 4]
```diff
@@ -1753,39 +1753,56 @@ ves_icall_System_Threading_Thread_SetPriority (MonoThreadObjectHandle this_obj,
...
+static MonoArrayHandle
+byte_array_to_domain (MonoArrayHandle arr, MonoDomain *domain, MonoError *error)
 {
-	MonoArray *copy;
+	HANDLE_FUNCTION_ENTER ()
 
-	error_init (error);
-	if (!arr)
-		return NULL;
+	if (!arr || MONO_HANDLE_IS_NULL (arr))
```
Comment (lambdageek) [inline mono/metadata/threads.c]: `!arr` should never happen.

File: mono/metadata/threads.c [hunk 5]
```diff
@@ -1753,39 +1753,56 @@ ves_icall_System_Threading_Thread_SetPriority (MonoThreadObjectHandle this_obj,
...
+	goto_if_nok (error, exit);
+
+	guint i;
+
+	// Pin both arrays.
+	for (i = 0; i <= 1; ++i)
+		arrays [i].p = mono_array_handle_pin_with_size (arrays [i].handle, size, 0, &arrays [i].gchandle);
+	goto_if_nok (error, exit);
+
+	memmove (arrays [1].p, arrays [0].p, size);
```
Comment (lambdageek) [inline mono/metadata/threads.c]: Maybe promote to a function like `mono_array_handle_memmove_norefs (MonoArrayHandle dest, guint32 dest_idx, MonoArrayHandle src, guint32 src_idx, guint32 nelems)` that does the pinning  and memmove?  Might be useful to have in the future.

Comment (jaykrell) [reply mono/metadata/threads.c]: Ugh I keep losing changes on this one and the array is back.

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64

**Comment by jaykrell** [comment]
@luhenry review or clear block?

**Pull request merged**